ir/function.cpp

```diff
@@ -289,4 +289,96 @@ void CFG::printDot(ostream &os) const {
   os << "}\n";
 }
 
+// Cooper, Keith D.; Harvey, Timothy J.; and Kennedy, Ken (2001). 
+// A Simple, Fast Dominance Algorithm
+// http://www.cs.rice.edu/~keith/EMBED/dom.pdf
+void DomTree::buildDominators() {
+  unsigned bb_count = f.getBBs().size();
+
+  // build predecessor relationships
+  for (const auto &[p, b, instr] : cfg) {
+    (void)instr;
+
+    auto &b_node = doms[&b];
+    if (!b_node.get())
+      b_node = make_unique<DomTreeNode>(b);
+    
+    auto &pred = doms[&p];
+    if (!pred.get())
+      pred = make_unique<DomTreeNode>(p);
+
+    b_node->preds.emplace_back(pred.get());
+  }
+
+  unsigned i = bb_count-1;
+  DomTreeNode *new_idom;
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (auto &b : f.getBBs()) {
```

Comment:
is this the algorithm of figure 3?
If so, it says "for all nodes, b, in reverse postorder (except start node)".  Do we guarantee that order?

Reply:
Yes, the reverse postorder is the topological order of f.getBBs(), both the article "Graph Algorithms with a Functional Flavour" from John Launchbury and a lecture from cornell (https://www.cs.cornell.edu/courses/cs2112/2012sp/lectures/lec24/lec24-12sp.html) say the same, and from what i tested i have not seen any differences.
The order attribute is then needed for intersect. The start node will be skipped as it has no predicates.
In 3hours when the llvm tests are finished i will push the new changes which includes a bugfix for loops.

Reply:
So you rely on top_sort() being run before this function is called.
This should be documented and the the code inspected to guarantee that's the case.